numpy/lib/_function_base_impl.py

```diff
@@ -530,16 +549,21 @@ def average(a, axis=None, weights=None, returned=False, *,
                 raise TypeError(
                     "Axis must be specified when shapes of a and weights "
                     "differ.")
-            if wgt.ndim != 1:
-                raise TypeError(
-                    "1D weights expected when shapes of a and weights differ.")
-            if wgt.shape[0] != a.shape[axis]:
+            if np.array(axis).shape[0] != wgt.ndim:
```

Comment:
I think the logic can be substantially easier:
```
# Keep the TypeError for wrong dimension for backward compatibility in tests.
if wgt.ndim != len(axis):
    raise TypeError(f"{len(axis)}D weights expected ... ")
# Check the shape.
if wgt.shape != tuple(a.shape[ax] for ax in axis):
    raise ValueError(...)
# Broadcast appropriately.
wgt = wgt.reshape(tuple(s if ax in axis else 1) for ax, s in enumerate(a.shape))
wgt = np.broadcast_to(wgt, a.shape, subok=True)
```
Note that the second line needs the `normalize_axis_tuple` from the earlier comment.

Also, the addition of `subok=True` ensures that `wgt` is not forced into a regular array (it should not, since we let subclasses in via `np.asanyarray`).

Reply:
Just to confirm, I guess we still need the check for `axis=None` (and raise a `TypeError`)? And then your 2 suggested checks. Thanks in advance.

Reply:
Another separate question.

In your implementation, `wgt` (e.g. initially `wgt.shape = (2,2)`) ends up having the of shape `a` (e.g. `wgt.shape = a.shape = (2,2,3)`). Whereas in my implementation (based on the original), the size of `wgt` in the additional dimension ends up as 1 (e.g.  `wgt.shape = (2,2,**1**)`)  (and we rely on the broadcasting in `np.multiply()` which is called after). 

Do the two implementations result in the same thing, e.g. in terms of memory use? Naively, I might think that having 'my' code allocate an array of shape (2,2,1) is better than having it allocating an array of shape (2,2,3), even if it results in the same thing in the internals of `np.multiply()` since a) a priori, I don't know what happens in `np.multiply()`, and b) when it's under my control, surely it's better to allocate less, rather than more, memory.

Thanks in advance for your help.

Reply:
And another, more important one.

In your implementation, the broadcasted weights do not change when `axis` is transposed (e.g. `(0,1)` -> `(1,0)`), and thus the result of `np.average()` does not change either. However that's not quite right, is it? 

The weights supplied should correspond to what you want to do along the supplied `axis` (at least, I think so), e.g. `weights[i]` apply to `axis[i]`. So transposing the axis should be equivalent to transposing the weights (I implemented a test for this). Do you agree?

My logic (based on the original) has this effect, and I'm not immediately sure how to adapt your simpler logic to account for this.

Reply:
@jamesoliverh - on broadcasting, I thought after I wrote too that the broadcasting would not be necessary. So, feel free to omit that.

On the axes, yes, you're right, inverted axes does not get treated correctly... Thankfully, you wrote good tests so this is obvious! 
So, one needs some swapping, but multiple calls to `swapaxes` is a bit inefficient as that just calls transpose and one can do all of them in a single transpose. I think just inserting the following before the reshape will do it (test routine below...):
```
wgt = wgt.transpose(np.argsort(axis))
```

Routine I used to check:
```
from numpy.core.numeric import normalize_axis_tuple

def reorder(a, w, axis):
     axis = normalize_axis_tuple(axis, a.ndim)
     w = w.transpose(np.argsort(axis))
     return w.reshape(tuple((s if ax in axis else 1) for ax, s in enumerate(a.shape)))

reorder(np.zeros((2, 3, 4, 5)), np.arange(12).reshape((4, 3)), axis=(2, 1))
array([[[[ 0],
         [ 3],
         [ 6],
         [ 9]],

        [[ 1],
         [ 4],
         [ 7],
         [10]],

        [[ 2],
         [ 5],
         [ 8],
         [11]]]])

reorder(np.zeros((2, 3, 4, 5)), np.arange(12).reshape((3, 4)), axis=(1, -2))
array([[[[ 0],
         [ 1],
         [ 2],
         [ 3]],

        [[ 4],
         [ 5],
         [ 6],
         [ 7]],

        [[ 8],
         [ 9],
         [10],
         [11]]]])
```

Reply:
And, yes, the idea was still to error on `axis is None` in this branch. For the other error, in astropy we are usually quite strict about not changing anything a downstream user might rely on, and this includes the type of error. But honestly, in this case, your single `ValueError` really is good enough. Also, since we are aiming for numpy 2.0, a bit of sanitizing is in fact allowed, and having separate errors for essentially the same thing makes little sense. So, just have the one!

Reply:
@mhvk Regarding the suggestion of `argsort()` for the transposition, that's fantastic, thank you. I've added that, and left out the broadcasting.

@mhvk Regarding the sanity checking and errors, thank you. One comment: for the ValueErrors I've opted for now to leave my original messages (stopping short of adding e.g. f"you need {len(axis)}D weights, dumbo"). I think my generic messages are clear as to what's expected and my natural inclination is to stop there, but happy to add dynamically generated information if you think it's necessary here.

Notwithstanding the comment, I think that should be everything - what do I need to do next?

Reply:
I see I have some issues to fix in the docs, will work on that (and/or ask for someone's help if I get stuck, if that's OK...)

Reply:
OK, all done, I would appreciate if someone could please let me know the next steps.